### PR TITLE
New ANT task with CC

### PR DESCRIPTION
**Pull request by prmittal**

### CODE REVIEW

File: Tasks/ANTPreview/task.json [hunk 1]
```diff
@@ -0,0 +1,168 @@
+{
+    "id": "18033022-9747-4b7b-8395-4fd7eb2b565a",
+    "name": "AntCC",
+    "friendlyName": "Ant With CodeCoverage",
```
Comment (allendm-msft) [inline Tasks/ANTPreview/task.json]: dont you want to mark the name with "preview" tag too? will get confusing for a user who enables preview tasks to see two tasks with same name?

File: Tasks/ANTPreview/task.json [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    "description": "Build with Apache Ant and check CodeCoverage",
+    "helpMarkDown": "[More Information](http://go.microsoft.com/fwlink/?LinkID=613718)",
+    "category": "Build",
+    "visibility": [
+                "Preview",
+                "Build"
+                  ],    
+    "author": "Microsoft Corporation",
+    "version": {
+        "Major": 1,
```
Comment (allendm-msft) [inline Tasks/ANTPreview/task.json]: shouldnt be a 1.0 version.. ideally 0.1.0

File: Tasks/ANTPreview/task.json [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+            "helpMarkDown": "Select the tool with which summary files have been generated",
+            "options": {
+                "JaCoCo": "JaCoCo"
+            }
+        },
+        {
+            "name": "summaryFileLocation",
+            "type": "string",
+            "label": "Summary File",
+            "defaultValue": "$(Build.Repository.LocalPath)\\my-app\\target\\site\\jacoco\\jacoco.xml",
```
Comment (allendm-msft) [inline Tasks/ANTPreview/task.json]: will it always be "my-app" or is it depedent on the app name?

File: Tasks/ANTPreview/task.json [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+            "helpMarkDown": "Select this option to publish Code Coverage results produced by the Maven build to VSO/TFS.Code Coverage Summary and code Coverage Files would be published to TFS Server"
+        },
+        {
+            "name": "codeCoverageTool",
+            "type": "pickList",
+            "label": "CodeCoverage Tool",
+            "required": false,
+            "groupName": "codeCoverage",
+            "defaultValue": "JaCoCo",
+            "helpMarkDown": "Select the tool with which summary files have been generated",
```
Comment (allendm-msft) [inline Tasks/ANTPreview/task.json]: I would call out that today we have only Jacoco support in the help itself. "Select the tool with which summary files have been generated. Only Jacoco support available currently"

File: Tasks/ANTPreview/task.json [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+            }
+        },
+        {
+            "name": "summaryFileLocation",
+            "type": "string",
+            "label": "Summary File",
+            "defaultValue": "$(Build.Repository.LocalPath)\\my-app\\target\\site\\jacoco\\jacoco.xml",
+            "required": true,
+            "groupName": "codeCoverage",
+            "helpMarkDown": "Path of codecoverage Summary file"
```
Comment (allendm-msft) [inline Tasks/ANTPreview/task.json]: small "s"

File: Tasks/ANTPreview/task.json [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+            "helpMarkDown": "Path of codecoverage report directory"
+        },
+        {
+            "name": "additionalCodeCoverageFiles",
+            "type": "filePath",
+            "label": "Additional files",
+            "defaultValue": "**/*.exec",
+            "required": true,
+            "groupName": "codeCoverage",
+            "helpMarkDown": "regular expression Specifying the additional codecoverage Files to be attached to build"
```
Comment (allendm-msft) [inline Tasks/ANTPreview/task.json]: small "s"

File: Tasks/ANTPreview/task.json [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+            "defaultValue": "$(Build.Repository.LocalPath)\\my-app\\target\\site\\jacoco",
+            "required": true,
+            "groupName": "codeCoverage",
+            "helpMarkDown": "Path of codecoverage report directory"
+        },
+        {
+            "name": "additionalCodeCoverageFiles",
+            "type": "filePath",
+            "label": "Additional files",
+            "defaultValue": "**/*.exec",
```
Comment (allendm-msft) [inline Tasks/ANTPreview/task.json]: is this the right "/" ? shouldnt it be "\" ?

File: Tasks/ANTPreview/ant.ps1 [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+else
+{
+    Write-Verbose "Option to publish JUnit Test results produced by Ant build was not selected and is being skipped."
+}
+
+
+if($publishCodeCoverageResults)
+{
+   # Publish Code Coverage Files
+   $CodeCoverageFiles = Find-Files -SearchPattern $additionalCodeCoverageFiles
```
Comment (allendm-msft) [inline Tasks/ANTPreview/ant.ps1]: follow naming conventions.. it should be $codeCoverageFiles.. small "c"

Comment (allendm-msft) [inline Tasks/ANTPreview/ant.ps1]: error conditions?? what happens if you dont find any files to publish?

File: Tasks/ANTPreview/task.json (hunk 1 above)
Comment (prmittal) [inline Tasks/ANTPreview/task.json]: The name for new tasks contains string "code coverage"

File: Tasks/ANTPreview/task.json (hunk 7 above)
Comment (prmittal) [inline Tasks/ANTPreview/task.json]: using the same as test results pattern

File: Tasks/ANTPreview/ant.ps1 (hunk 8 above)
Comment (prmittal) [inline Tasks/ANTPreview/ant.ps1]: i dont think it should be either. its okay if no files are found.

Comment (allendm-msft) [inline Tasks/ANTPreview/ant.ps1]: It cant be neither. The user checked the publishCC checkbox and is expecting the results and you will not inform him/her that it wasnt generated?
Add in a warning message

File: Tasks/ANTPreview/task.json [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+    "description": "Build with Apache Ant and check CodeCoverage",
+    "helpMarkDown": "[More Information](http://go.microsoft.com/fwlink/?LinkID=613718)",
+    "category": "Build",
+    "visibility": [
+                "Preview",
+                "Build"
+                  ],    
+    "author": "Microsoft Corporation",
+    "version": {
+        "Major": 0,
```
Comment (ajay-arora) [inline Tasks/ANTPreview/task.json]: usually we start with version 1.

File: Tasks/ANTPreview/task.json [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+            "helpMarkDown": "Select the tool with which summary files have been generated. Only Jacoco support available currently. Jacocoant.jar should be present in lib folder of Ant installation.",
+            "options": {
+                "JaCoCo": "JaCoCo"
+            }
+        },
+        {
+            "name": "summaryFileLocation",
+            "type": "string",
+            "label": "Summary File",
+            "defaultValue": "$(Build.Repository.LocalPath)\\target\\site\\jacoco\\jacoco.xml",
```
Comment (ajay-arora) [inline Tasks/ANTPreview/task.json]: there can be more than 1 code coverage configured inside ant build file ... how are we planning to address that ?

File: Tasks/ANTPreview/ant.ps1 [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+   Publish-CodeCoverage -CodeCoverageTool $codeCoverageTool -SummaryFileLocation $summaryFileLocation -ReportDirectory $reportDirectory -AdditionalCodeCoverageFiles $codeCoverageFiles -Context $distributedTaskContext    
+}
+else
+{
+    Write-Verbose "Option to publish CodeCoverage results produced by Maven build was not selected and is being skipped."
+}
+
+
+Write-Verbose "Leaving script Ant.ps1"
+
```
Comment (ajay-arora) [inline Tasks/ANTPreview/ant.ps1]: last few lines are empty .. pls remove them

File: Tasks/ANTPreview/ant2.js [hunk 12]
```diff
@@ -0,0 +1,74 @@
+var tl = require('vso-task-lib');
```
Comment (ajay-arora) [inline Tasks/ANTPreview/ant2.js]: not clear on why do we need to clone the js as well if we are not making any changes.
we would have to worry about taking FI from the parent branch of ANT.

**Pull request merged**